core/src/main/scala/org/apache/spark/internal/io/HadoopMapReduceCommitProtocol.scala

```diff
@@ -271,4 +318,59 @@ class HadoopMapReduceCommitProtocol(
         logWarning(s"Exception while aborting ${taskContext.getTaskAttemptID}", e)
     }
   }
+
+  /**
+   * Delete the staging output path of current InsertIntoHadoopFsRelation operation.
+   */
+  private def deleteStagingInsertOutputPath(fs: FileSystem): Unit = {
+    if (staticPartitionKVs.size == 0) {
+      fs.delete(insertStagingDir, true)
+    } else {
+      var currentLevelPath = new Path(insertStagingDir, getStaticPartitionPath(staticPartitionKVs))
+      fs.delete(currentLevelPath, true)
+
+      var complete = false
+      var remainingLevel = staticPartitionKVs.size - 1
+      while (!complete && remainingLevel > 0) {
+        try {
+          currentLevelPath = new Path(insertStagingDir,
+            getStaticPartitionPath(staticPartitionKVs.slice(0, remainingLevel)))
+          if (!fs.delete(currentLevelPath, false)) {
+            complete = true
+          }
+          remainingLevel -= 1
+        } catch {
+          case e: Exception =>
+            logWarning(s"Exception occurred when deleting dir: $currentLevelPath.", e)
+            complete = true
+        }
+      }
+
+      try {
+        fs.delete(insertStagingDir, false)
+      } catch {
+        case e: Exception =>
+          logWarning(s"Exception occurred when deleting dir: $insertStagingDir.", e)
+      }
+    }
+  }
+}
+
+object  HadoopMapReduceCommitProtocol {
+
+  private def escapePathName(path: String): String = {
```

Comment:
Why doing this way?  Because we are in the core package?

Do we have better alternatives? How about we pre escapePathname early, like before passing to FileCommitProtocol?

Reply:
Yes, we are in core package and  can not import this class.